lib/error_report_notify.c

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * Userspace notification interface for debugging tools.
+ *
+ * Provide two sysfs files:
+ *  - /sys/kernel/error_report/last_report
+ *  - /sys/kernel/error_report/report_count
+ * that contain the last debugging tool report (taken from dmesg, delimited by
+ * the error_report_start/error_report_end tracing events) and the total report
+ * count.
+ *
+ * Userspace apps can call poll(POLLPRI) on those files to get notified about
+ * the new reports without having to watch dmesg in a loop.
+ *
+ */
+
+#include <linux/fs.h>
+#include <linux/string.h>
+#include <linux/tracepoint.h>
+#include <linux/workqueue.h>
+#include <trace/events/error_report.h>
+#include <trace/events/printk.h>
+
+static struct kobject *error_report_kobj;
+static atomic_t num_error_reports;
+
+static char current_report[PAGE_SIZE], last_report[PAGE_SIZE];
+static size_t current_report_pos, last_report_size;
+static atomic_t current_pid = ATOMIC_INIT(-1);
+/* Protects last_report and last_report_size. */
+static DEFINE_SPINLOCK(last_report_lock);
+
+static struct delayed_work reporting_done;
+
+static void error_report_notify(struct work_struct *work)
+{
+	sysfs_notify(error_report_kobj, NULL, "last_report");
+	sysfs_notify(error_report_kobj, NULL, "report_count");
+}
+static DECLARE_DELAYED_WORK(reporting_done, error_report_notify);
+
+/*
+ * Trace hook for the error_report_start event. If two reports overlap (the
+ * second one report is started before the first one is finished, the first
+ * report is discarded.
+ */
+static void probe_report_start(void *ignore, const char *tool_name,
+			       unsigned long id)
+{
+	atomic_set_release(&current_pid, current->pid);
```

Comment:
This should be a cmpxchg:
```
/* Try to become exclusive error reporter; any concurrent errors will be discarded. */
atomic_cmpxchg_relaxed(&current_pid, -1, current->pid);
```

But that doesn't quite work, because we might have a nested task --> interrupt case. If the interrupt reports an error it's going to be a mess, because while it won't succeed in the cmpxchg, it'll start recording info.

I think we should guard against that case, but can probably say if we get nested interrupt a'la task -> interrupt --> interrupt/nmi, then all bets are off and we don't care.

```
/* <explaining what I said above> */
static int get_encoded_pid(void) { return (current->pid << 1) | (!!in_task()); }
```

And then use get_encoded_pid() instead of current->pid throughput. That way nested task -> interrupt will work just fine and the interrupt will fail to cmpxchg, and will also not start recording console output.